state/protocol/protocol_state/always_evolvestate_wrapper.go

```diff
@@ -0,0 +1,50 @@
+package protocol_state
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol"
+)
+
+// AlwaysEvolveStateWrapper wraps a `protocol_state.OrthogonalStoreStateMachine`. The wrapper guarantees that
+// the state machine's `EvolveState(…)` is always called before its `Build()` method. If it hasn't been executed
+// before, the wrapper will call the state machine's `EvolveState(…)` method first in its `Build()` step.
+//
+// NOT CONCURRENCY SAFE
+type AlwaysEvolveStateWrapper[P any] struct {
+	OrthogonalStoreStateMachine[P]
+	evolveStateCalled bool
+}
```

Comment:
I see the reasoning here, but I'm not a fan of this approach. In particular, I am worried about the fallback mechanism (where `Build` is called when we have not yet called `EvolveState`) silently passing an empty service event list. In cases where the logic is extended without following the interface requirements (what we were worried about to begin with), this can just as easily cause failures. 

In general I feel like this component is adding complexity but not gaining meaningful safety improvements.

Reply:
Thanks for your feedback. You argument resonates with me.

#### Lets recap:

_Challenge at hand_:
* Relatively frequently, there are no seals in a block or the sealed results do not contain ay Service Events. However, we want to allow State Machines to utilize view-based triggers that can change the protocol state, irrespective whether there are Service Events in the block. 

_Possible Approaches_:
* (a) every `OrthogonalStoreStateMachine` must handle the case that `EvolveState` _may_ not be called if there are no service events (or it may be called with an empty list) before `Build`
   * The Dynamic Protocol State is a framework for adding protocol-relevant values and evolving them via `OrthogonalStoreStateMachine`s. Requiring every engineer that adds a new `OrthogonalStoreStateMachine` to handle this same edge case seems risky. I think option (a) is acceptable, but certainly not my preference.
   * The `AlwaysEvolveStateWrapper` falls in this category, because it merely adds the `AlwaysEvolveStateWrapper` as a tool to help address our challenge. Nevertheless, engineers adding `OrthogonalStoreStateMachine` still need to be aware of this edge case and explicitly utilize `AlwaysEvolveStateWrapper` for their state machine. 

The other alternative is broadly that the Dynamic Protocol State guarantees that  `EvolveState` is _always_ called before `Build()`. I think we can add this guarantee at different levels:
* (b) the [`stateMutator`](https://github.com/onflow/flow-go/blob/6b5dcaf306ebe0cf6e613b5c272f71f8b5d96c8a/state/protocol/protocol_state/state/mutator.go#L28) guarantees to execute `ApplyServiceEventsFromValidatedSeals` with an empty list of seals as the first step in `Build` if and only if `ApplyServiceEventsFromValidatedSeals` hasn't been executed before
* (c) we offload the responsibility to _always_ call [`StateMutator.ApplyServiceEventsFromValidatedSeals(..)`](https://github.com/onflow/flow-go/blob/a203323be24d95dde54187559282d9b5f6dea667/state/protocol/protocol_state.go#L178) to the higher-level logic. In particular, the [`consensus.Builder`](https://github.com/onflow/flow-go/blob/6b5dcaf306ebe0cf6e613b5c272f71f8b5d96c8a/module/builder/consensus/builder.go#L642-L646) as well as the [`FollowerState`](https://github.com/onflow/flow-go/blob/6b5dcaf306ebe0cf6e613b5c272f71f8b5d96c8a/state/protocol/badger/mutator.go#L600-L614) then _must_ implement this guarantee
  * in fact, both implementations currently do the correct thing
  * But, the requirement is currently not documented and therefore easily broken (and probably also with proper documentation easily overlooked). But we could add dedicated unit tests to help avoid breaking this guarantee accidentally in the future. 
  * Nevertheless, by offloading the requirement to call [`StateMutator.ApplyServiceEventsFromValidatedSeals(..)`](https://github.com/onflow/flow-go/blob/a203323be24d95dde54187559282d9b5f6dea667/state/protocol/protocol_state.go#L178) to the higher-level logic, we have more than one place in the business logic that needs to be aware of this requirement 

Curious about your opinions. I am currently leaning towards option (b). Totally  happy to remove `AlwaysEvolveStateWrapper` and revert the associated changed, but wanted to get your input first.

Reply:
#### quick summary of conversation with Yurii and Jordan:
for this PR, we decided to move ahead with a variant of option (c):
   * rename [StateMutator.ApplyServiceEventsFromValidatedSeals(..)](https://github.com/onflow/flow-go/blob/a203323be24d95dde54187559282d9b5f6dea667/state/protocol/protocol_state.go#L178) to `EvolveState(..)` for clarity that this step is not only about service events
   * I'll extend the API documentation to explicitly emphasize the need that `EvolveState(..)` (formerly `ApplyServiceEventsFromValidatedSeals`) is called
   * add unit tests to [consensus.Builder](https://github.com/onflow/flow-go/blob/6b5dcaf306ebe0cf6e613b5c272f71f8b5d96c8a/module/builder/consensus/builder.go#L642-L646) as well as the [FollowerState](https://github.com/onflow/flow-go/blob/6b5dcaf306ebe0cf6e613b5c272f71f8b5d96c8a/state/protocol/badger/mutator.go#L600-L614) to enforce that they call `EvolveState(..)` (formerly `ApplyServiceEventsFromValidatedSeals`). Thereby, we guarantee that each individual state machine's `EvolveState(..)` method is always called. 
  
  The [consensus.Builder](https://github.com/onflow/flow-go/blob/6b5dcaf306ebe0cf6e613b5c272f71f8b5d96c8a/module/builder/consensus/builder.go#L642-L646) and  [FollowerState](https://github.com/onflow/flow-go/blob/6b5dcaf306ebe0cf6e613b5c272f71f8b5d96c8a/state/protocol/badger/mutator.go#L600-L614) are relatively mature and infrequently-changing code, which will only be modified by engineers that are very familiar with the respective protocol logic. The logic for evolving the protocol state is very straight forward in the [consensus.Builder](https://github.com/onflow/flow-go/blob/6b5dcaf306ebe0cf6e613b5c272f71f8b5d96c8a/module/builder/consensus/builder.go#L642-L646) and  [FollowerState](https://github.com/onflow/flow-go/blob/6b5dcaf306ebe0cf6e613b5c272f71f8b5d96c8a/state/protocol/badger/mutator.go#L600-L614), so complexity increase in negligible. Therefore, we concluded that it is the safest option to implement the requirement to call `EvolveState(..)` in [consensus.Builder](https://github.com/onflow/flow-go/blob/6b5dcaf306ebe0cf6e613b5c272f71f8b5d96c8a/module/builder/consensus/builder.go#L642-L646) and  [FollowerState](https://github.com/onflow/flow-go/blob/6b5dcaf306ebe0cf6e613b5c272f71f8b5d96c8a/state/protocol/badger/mutator.go#L600-L614)

Reply:
`ApplyServiceEventsFromValidatedSeals` was removed and I consolidated the logic into `EvolveState` (see PR description above for a more extensive overview of changes)